software/contrib/turing_machine.py

```diff
@@ -0,0 +1,254 @@
+    import europi
+    from europi import clamp, din, ain, k1, k2, b1, b2, cv1, cv2, cv3, cv4, cv5, cv6, oled
+
+from europi_script import EuroPiScript
+
+# Customize pulses output here
+CV1_PULSE_BIT = 1
+CV2_PULSE_BIT = 2
+CV3_PULSE_BIT = 4
+
+INT_MAX_8 = 0xFF
+DEFAULT_BIT_COUNT = 16
+MAX_OUTPUT_VOLTAGE = europi.MAX_OUTPUT_VOLTAGE
+
+
+class TuringMachine:
+    """A class meant to recreate the Music Thing Modular Turning Machine Random Sequencer as faithfully as possible in
+    micropython using bit shift operations to mimic the TM's analog shift register.
+
+    The ``TuringMachine`` class keeps its state in several internal member variables which are accessed via property
+    variables. See ``flip_probability``, ``scale``, ``length`` and ``write``. users of the class can modify these
+    variables in order to configure the TM's behavior. In many cases script authors will want to tie one of these
+    variables to a hardware UI element, for example, assigning the ``flip_probability`` to a knob. In this case clients
+    can override the corresponding ``*_getter`` function. For example, to tie the write switch to button 1::
+
+        tm = TuringMachine()
+
+        def write(self):
+          return b1.value()
+
+        tm.write_getter = write
+
+    This form allows clients to override where the ``TuringMachine`` obtains some of its state variables without
+    explicitly sub-classing the ``TuringMachine`` class. See ``EuroPiTuringMachine`` for a more detailed example.
+    """
+
+    def __init__(self, bit_count=DEFAULT_BIT_COUNT, max_output_voltage=MAX_OUTPUT_VOLTAGE):
+        """Create a new TuringMachine with a shift register of the specified bit count. Default is 16, minimum is 8.
+        The maximum output voltage is also configurable and defaults to ``europi.MAX_OUTPUT_VOLTAGE``"""
+        if bit_count < 8:
+            raise ValueError(f"Specified bit_count ({bit_count}) is less than the minimum (8).")
+        self.bit_count = bit_count
+        self.bits = getrandbits(self.bit_count)
+        self._flip_probability = 0
+        self.max_output_voltage = max_output_voltage
+        self._scale = max_output_voltage
+        self._length = bit_count
+        self._write = False
+
+        self.flip_probability_getter = lambda: self._flip_probability
+        self.scale_getter = lambda: self._scale
+        self.length_getter = lambda: self._length
+        self.write_getter = lambda: self._write
+        self.step_handler = lambda: None
+
+    def _rotate_bits(self):
+        self.bits = ((self.bits << 1) % (1 << self.bit_count)) | ((self.bits >> (self.length - 1)) & 1)
+
+    def step(self):
+        """Move the turing machine to its next state based on its current state. Parameters that affect the next state
+        include: ``flip_probability``, ``length``, ``write``, and, the internal bit register.
+
+        Typically this method would be called in response to a clock tick.
+        """
+        self._rotate_bits()
```

Comment:
My bit-wise skills are rusty, so I'll have to trust these all work as intended (and the tests help provide confidence there)!

Reply:
I've always stayed away from bit work as well. I took this as an opportunity to brush up on it. And yes, I couldn't have done it without writing those tests.